File: app/src/main/java/com/woaigmz/result/ActivityResultRegistry.java
package com.woaigmz.result;

import android.app.Activity;
import android.content.Intent;
import android.util.SparseArray;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.LifecycleEventObserver;
import androidx.lifecycle.LifecycleOwner;

import java.util.concurrent.atomic.AtomicInteger;

/**
 * Created by haoran on 2020/4/22.
 */
public abstract class ActivityResultRegistry {

    private SparseArray<ActivityResultCallback> requestCodeToCallback = new SparseArray<>();
    private final AtomicInteger code = new AtomicInteger(0);


    <O> ActivityResultLauncher register(LifecycleOwner owner, ActivityResultCallback<O> callback){
        final int requestCode = this.code.getAndIncrement();
        requestCodeToCallback.put(requestCode,callback);
        Lifecycle lifecycle = owner.getLifecycle();
        lifecycle.addObserver(new LifecycleEventObserver() {
            @Override
            public void onStateChanged(@NonNull LifecycleOwner lifecycleOwner,
                                       @NonNull Lifecycle.Event event) {
                if (Lifecycle.Event.ON_DESTROY.equals(event)) {
                    unregister(requestCode);
                }
            }
        });

        return new ActivityResultLauncher() {
            @Override
            public void launch(Intent intent) {
                invoke(requestCode,intent);
            }
        };
    }

    abstract void invoke(int rc, Intent intent);


    private void unregister(int requestCode) {
        requestCodeToCallback.remove(requestCode);

    }

    public final boolean dispatchResult(int requestCode, int resultCode, Intent data) {
        ActivityResultCallback callback = requestCodeToCallback.get(requestCode);
        if(callback == null){
            return false;
        }
        doDispatch(resultCode,data,callback);
        return true;
    }

    private <O> void doDispatch(int resultCode, @Nullable Intent data,
                                @Nullable ActivityResultCallback<O> callback) {
        if (callback!=null){
            if(resultCode == Activity.RESULT_OK){
                callback.onGetResult(callback.parseResult(data));
            }else {
                callback.onGetResult(null);
            }
        }



    }
}

File: app/src/main/java/com/woaigmz/result/ActivityResultRegistry.kt
package com.woaigmz.result

import android.app.Activity
import android.content.Intent
import android.util.SparseArray
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleEventObserver
import androidx.lifecycle.LifecycleOwner
import com.woaigmz.result.Constants.OUTPUT_EXTRA
import java.util.concurrent.atomic.AtomicInteger

/**
 * Created by haoran on 2020/4/22.
 */

abstract class ActivityResultRegistry {

    private val requestCodeToCallback = SparseArray<ActivityResultCallback<*>>()
    private val code = AtomicInteger(0)

    fun <O> register(
        owner: LifecycleOwner,
        callback: ActivityResultCallback<O>
    ): ActivityResultLauncher {
        val requestCode = code.getAndIncrement()
        requestCodeToCallback.put(requestCode, callback)
        val lifecycle = owner.lifecycle
        lifecycle.addObserver(LifecycleEventObserver { _, event ->
            if (Lifecycle.Event.ON_DESTROY == event) {
                unregister(requestCode)
            }
        })
        return ActivityResultLauncher {
               intent -> invoke(requestCode, intent)
        }
    }

    abstract operator fun invoke(rc: Int, intent: Intent?)

    private fun unregister(requestCode: Int) {
        requestCodeToCallback.remove(requestCode)
    }

    fun dispatchResult(requestCode: Int, resultCode: Int, data: Intent?): Boolean {
        val callback = requestCodeToCallback[requestCode] ?: return false
        doDispatch(resultCode, data, callback)
        return true
    }

    @Suppress("UNCHECKED_CAST")
    private fun <O> doDispatch(
        resultCode: Int, data: Intent?,
        callback: ActivityResultCallback<O?>?
    ) {
        if (resultCode == Activity.RESULT_OK)
            callback?.onGetResult(data?.extras?.get(OUTPUT_EXTRA) as O?)
    }
}